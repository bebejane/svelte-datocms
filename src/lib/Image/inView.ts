type inViewOptions = {
	once: boolean;
	root: null | HTMLElement;
	rootMargin: string;
	threshold: number;
};
type ActionReturnType = {
	update: (options: inViewOptions) => void;
	destroy: () => void;
};

export default function (
	node: HTMLElement,
	options: inViewOptions = {
		threshold: 0,
		rootMargin: '0px',
		root: null,
		once: false
	}
): ActionReturnType {
    let observer : IntersectionObserver
	const dispatch = (isInView:boolean) => {
		const event = new CustomEvent<{isInView:boolean}>('inview', { detail: { isInView } });
		node.dispatchEvent(event);
	};
	const observerCallback: IntersectionObserverCallback = (entries) => {
		entries.forEach((entry) => {
			if (entry !== null) {
				dispatch( false);
				if (entry.isIntersecting) {
					dispatch(true);
					if (options.once) observer.unobserve(node);
				}
			}
		});
	};
    const initialize = ({root,rootMargin,threshold} = options)=>{
        observer = new IntersectionObserver(observerCallback, { root, rootMargin, threshold });
        observer.observe(node);
    }
    initialize()
	return {
		update(properties) {
            observer.disconnect()
            initialize(properties)
			console.log('update called!');
		},
		destroy() {
			observer && observer.disconnect();
		}
	};
}
