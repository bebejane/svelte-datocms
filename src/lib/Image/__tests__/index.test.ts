/**
 * @jest-environment jsdom
 */
 jest.mock('../inView')
 import inView from '../inView'
 import {mockIntersectionObserver} from 'jsdom-testing-mocks'
import { render, cleanup, fireEvent } from '@testing-library/svelte';
import { Image } from '../../index';

const data = {
	alt: 'DatoCMS swag',
	aspectRatio: 1.7777777777777777,
	base64:
		'data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z',
	height: 421,
	sizes: '(max-width: 750px) 100vw, 750px',
	src: 'https://www.datocms-assets.com/205/image.png?ar=16%3A9&fit=crop&w=750',
	srcSet:
		'https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.25&fit=crop&w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.5&fit=crop&w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.75&fit=crop&w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1&fit=crop&w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1.5&fit=crop&w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=2&fit=crop&w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=3&fit=crop&w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=4&fit=crop&w=750 3000w',
	title: 'These are awesome, we know that.',
	width: 750
};

mockIntersectionObserver()


describe('Image', () => {
	// intersectionThreshold is an hack to make tests work
	// we need the library to generate a different IntersectionObserver for each test
	// otherwise the IntersectionObserver mocking won't work
	afterEach(()=>cleanup())
	describe('not visible', () => {
		test('renders the blur-up thumb', () => {
			const { container } = render(Image, { data });
			expect(inView).toHaveBeenCalled();
			expect(container).toMatchSnapshot();

		});
	});
	describe('Visible', () => {
		test('renders the Image',async () => {
			const { container } = render(Image, { data });
			const rootImage = container.querySelector('.relative-positioning')
			const event = new CustomEvent<{isInView:boolean}>('inview', { detail: { isInView:true } });
			await fireEvent(rootImage, event)
			expect(container).toMatchSnapshot();
		});
	});
	describe('Image loaded', () => {
		test('shows the Image',async () => {
			const { container } = render(Image, { data });
			const rootImage = container.querySelector('.relative-positioning')
			const event = new CustomEvent<{isInView:boolean}>('inview', { detail: { isInView:true } });
			await fireEvent(rootImage, event)
			const image = container.querySelector('img.absolute-positioning')
			await fireEvent.load(image)
			expect(container).toMatchSnapshot();
		});
	});
	
});
