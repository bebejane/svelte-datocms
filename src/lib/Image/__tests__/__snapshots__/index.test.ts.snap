// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Image loaded shows the Image 1`] = `
<body>
  <div>
    <div
      class="relative-positioning svelte-5iyfh4"
      style="display:block"
    >
      <img
        alt="placeholder"
        class="d-block svelte-5iyfh4"
        role="presentation"
        src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI3NTAiIGhlaWdodD0iNDIxIj48L3N2Zz4="
        style="width:100%"
      />
       
      <div
        class="absolute-positioning svelte-5iyfh4"
        style="background-color:#000;background-image:url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z);background-size:cover;"
      />
       
      <picture>
        <source />
         
        <source
          srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.25&fit=crop&w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.5&fit=crop&w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.75&fit=crop&w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1&fit=crop&w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1.5&fit=crop&w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=2&fit=crop&w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=3&fit=crop&w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=4&fit=crop&w=750 3000w"
        />
         
        <img
          alt="DatoCMS swag"
          class="absolute-positioning transition-opacity svelte-5iyfh4 visible"
          src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&fit=crop&w=750"
          style="transition-duration:500ms;"
          title="These are awesome, we know that."
        />
      </picture>
    </div>
  </div>
</body>
`;

exports[`Image Visible renders the Image 1`] = `
<body>
  <div>
    <div
      class="relative-positioning svelte-5iyfh4"
      style="display:block"
    >
      <img
        alt="placeholder"
        class="d-block svelte-5iyfh4"
        role="presentation"
        src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI3NTAiIGhlaWdodD0iNDIxIj48L3N2Zz4="
        style="width:100%"
      />
       
      <div
        class="absolute-positioning svelte-5iyfh4 visible"
        style="background-color:#000;background-image:url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z);background-size:cover;"
      />
       
      <picture>
        <source />
         
        <source
          srcset="https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.25&fit=crop&w=750 187w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.5&fit=crop&w=750 375w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=0.75&fit=crop&w=750 562w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1&fit=crop&w=750 750w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=1.5&fit=crop&w=750 1125w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=2&fit=crop&w=750 1500w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=3&fit=crop&w=750 2250w,↵https://www.datocms-assets.com/205/image.png?ar=16%3A9&dpr=4&fit=crop&w=750 3000w"
        />
         
        <img
          alt="DatoCMS swag"
          class="absolute-positioning transition-opacity svelte-5iyfh4 invisible"
          src="https://www.datocms-assets.com/205/image.png?ar=16%3A9&fit=crop&w=750"
          style="transition-duration:500ms;"
          title="These are awesome, we know that."
        />
      </picture>
    </div>
  </div>
</body>
`;

exports[`Image not visible renders the blur-up thumb 1`] = `
<body>
  <div>
    <div
      class="relative-positioning svelte-5iyfh4"
      style="display:block"
    >
      <img
        alt="placeholder"
        class="d-block svelte-5iyfh4"
        role="presentation"
        src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI3NTAiIGhlaWdodD0iNDIxIj48L3N2Zz4="
        style="width:100%"
      />
       
      <div
        class="absolute-positioning svelte-5iyfh4 visible"
        style="background-color:#000;background-image:url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHBwgHBgoICAgLFQoLDhgQDg0NDh0eHREYIx8lJCIrHB0dLSs7GikyKSEuKjUlKDk1MjIyHyo4PTc+PDcxPjUBCgsLDg0OHBAQHDsoIig7Ozs7Ozs7OzsvOzs7Ozs7Ozs7Lzs7Ozs7Ozs7OzsvOzs7NTsvLy87NTU1Ly8vLzsvL//AABEIAA0AGAMBIgACEQEDEQH/xAAYAAACAwAAAAAAAAAAAAAAAAAGBwABBP/EACEQAAEEAAYDAAAAAAAAAAAAAAEAAgMEBQYHESEiFWFx/8QAFQEBAQAAAAAAAAAAAAAAAAAAAwL/xAAZEQADAAMAAAAAAAAAAAAAAAAAAQIRITH/2gAMAwEAAhEDEQA/AFxLgDWTsAd1J5TGy7hEYqNAaNgECX7sjLMQAHJTEy1Zcarfia4lJMauAxqBhLY6ZlaOzDurWvUOd3jZPfCiEh4xs//Z);background-size:cover;"
      />
       
    </div>
  </div>
</body>
`;
